In [576]:
import numpy as np
import pandas as pd
df  = pd.read_csv("C://Users//NIRBHAY IMI-K//Desktop//ML assignment 2//Steel_industry_data.csv")

In [577]:
df.head()

date  Usage_kWh  Lagging_Current_Reactive.Power_kVarh  \
0  01/01/2018 00:15       3.17                                  2.95   
1  01/01/2018 00:30       4.00                                  4.46   
2  01/01/2018 00:45       3.24                                  3.28   
3  01/01/2018 01:00       3.31                                  3.56   
4  01/01/2018 01:15       3.82                                  4.50   

   Leading_Current_Reactive_Power_kVarh  CO2(tCO2)  \
0                                   0.0        0.0   
1                                   0.0        0.0   
2                                   0.0        0.0   
3                                   0.0        0.0   
4                                   0.0        0.0   

   Lagging_Current_Power_Factor  Leading_Current_Power_Factor   NSM  \
0                         73.21                         100.0   900   
1                         66.77                         100.0  1800   
2                         70.28                         100.0  2700   
3                         68.09                         100.0  3600   
4                         64.72                         100.0  4500   

  WeekStatus Day_of_week   Load_Type  
0    Weekday      Monday  Light_Load  
1    Weekday      Monday  Light_Load  
2    Weekday      Monday  Light_Load  
3    Weekday      Monday  Light_Load  
4    Weekday      Monday  Light_Load

In [578]:
df.shape

(35040, 11)

In [579]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35040 entries, 0 to 35039
Data columns (total 11 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   date                                  35040 non-null  object 
 1   Usage_kWh                             35040 non-null  float64
 2   Lagging_Current_Reactive.Power_kVarh  35040 non-null  float64
 3   Leading_Current_Reactive_Power_kVarh  35040 non-null  float64
 4   CO2(tCO2)                             35040 non-null  float64
 5   Lagging_Current_Power_Factor          35040 non-null  float64
 6   Leading_Current_Power_Factor          35040 non-null  float64
 7   NSM                                   35040 non-null  int64  
 8   WeekStatus                            35040 non-null  object 
 9   Day_of_week                           35040 non-null  object 
 10  Load_Type                             35040 non-null  object 
dtypes: float64(6), 

In [580]:
df.isna().sum()

date                                    0
Usage_kWh                               0
Lagging_Current_Reactive.Power_kVarh    0
Leading_Current_Reactive_Power_kVarh    0
CO2(tCO2)                               0
Lagging_Current_Power_Factor            0
Leading_Current_Power_Factor            0
NSM                                     0
WeekStatus                              0
Day_of_week                             0
Load_Type                               0
dtype: int64

In [581]:
df.drop(["date"], axis=1,inplace=True)

In [582]:
df.head()

Usage_kWh  Lagging_Current_Reactive.Power_kVarh  \
0       3.17                                  2.95   
1       4.00                                  4.46   
2       3.24                                  3.28   
3       3.31                                  3.56   
4       3.82                                  4.50   

   Leading_Current_Reactive_Power_kVarh  CO2(tCO2)  \
0                                   0.0        0.0   
1                                   0.0        0.0   
2                                   0.0        0.0   
3                                   0.0        0.0   
4                                   0.0        0.0   

   Lagging_Current_Power_Factor  Leading_Current_Power_Factor   NSM  \
0                         73.21                         100.0   900   
1                         66.77                         100.0  1800   
2                         70.28                         100.0  2700   
3                         68.09                         100.0  3600   
4                         64.72                         100.0  4500   

  WeekStatus Day_of_week   Load_Type  
0    Weekday      Monday  Light_Load  
1    Weekday      Monday  Light_Load  
2    Weekday      Monday  Light_Load  
3    Weekday      Monday  Light_Load  
4    Weekday      Monday  Light_Load

In [583]:
df.describe()

Usage_kWh  Lagging_Current_Reactive.Power_kVarh  \
count  35040.000000                          35040.000000   
mean      27.386892                             13.035384   
std       33.444380                             16.306000   
min        0.000000                              0.000000   
25%        3.200000                              2.300000   
50%        4.570000                              5.000000   
75%       51.237500                             22.640000   
max      157.180000                             96.910000   

       Leading_Current_Reactive_Power_kVarh     CO2(tCO2)  \
count                          35040.000000  35040.000000   
mean                               3.870949      0.011524   
std                                7.424463      0.016151   
min                                0.000000      0.000000   
25%                                0.000000      0.000000   
50%                                0.000000      0.000000   
75%                                2.090000      0.020000   
max                               27.760000      0.070000   

       Lagging_Current_Power_Factor  Leading_Current_Power_Factor  \
count                  35040.000000                  35040.000000   
mean                      80.578056                     84.367870   
std                       18.921322                     30.456535   
min                        0.000000                      0.000000   
25%                       63.320000                     99.700000   
50%                       87.960000                    100.000000   
75%                       99.022500                    100.000000   
max                      100.000000                    100.000000   

                NSM  
count  35040.000000  
mean   42750.000000  
std    24940.534317  
min        0.000000  
25%    21375.000000  
50%    42750.000000  
75%    64125.000000  
max    85500.000000

In [584]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [585]:
df.columns

Index(['Usage_kWh', 'Lagging_Current_Reactive.Power_kVarh',
       'Leading_Current_Reactive_Power_kVarh', 'CO2(tCO2)',
       'Lagging_Current_Power_Factor', 'Leading_Current_Power_Factor', 'NSM',
       'WeekStatus', 'Day_of_week', 'Load_Type'],
      dtype='object')

In [586]:
df['WeekStatus']=le.fit_transform(df['WeekStatus'])
df['Day_of_week']=le.fit_transform(df['Day_of_week'])
df['Load_Type']=le.fit_transform(df['Load_Type'])

In [587]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

In [588]:
cols = ['Usage_kWh', 'Lagging_Current_Reactive.Power_kVarh',
       'Leading_Current_Reactive_Power_kVarh', 'CO2(tCO2)',
       'Lagging_Current_Power_Factor', 'Leading_Current_Power_Factor', 'NSM']

In [589]:
df[cols]=ss.fit_transform(df[cols])

In [590]:
df.head()

Usage_kWh  Lagging_Current_Reactive.Power_kVarh  \
0  -0.724105                             -0.618516   
1  -0.699287                             -0.525911   
2  -0.722012                             -0.598278   
3  -0.719919                             -0.581106   
4  -0.704669                             -0.523458   

   Leading_Current_Reactive_Power_kVarh  CO2(tCO2)  \
0                             -0.521385   -0.71355   
1                             -0.521385   -0.71355   
2                             -0.521385   -0.71355   
3                             -0.521385   -0.71355   
4                             -0.521385   -0.71355   

   Lagging_Current_Power_Factor  Leading_Current_Power_Factor       NSM  \
0                     -0.389410                      0.513268 -1.678015   
1                     -0.729772                      0.513268 -1.641929   
2                     -0.544264                      0.513268 -1.605843   
3                     -0.660009                      0.513268 -1.569756   
4                     -0.838117                      0.513268 -1.533670   

   WeekStatus  Day_of_week  Load_Type  
0           0            1          0  
1           0            1          0  
2           0            1          0  
3           0            1          0  
4           0            1          0

In [591]:
from sklearn.model_selection import train_test_split

In [592]:
X = df[['Lagging_Current_Reactive.Power_kVarh',
       'Leading_Current_Reactive_Power_kVarh', 'CO2(tCO2)',
       'Lagging_Current_Power_Factor', 'Leading_Current_Power_Factor', 'NSM']]
Y = df['Usage_kWh']

In [593]:
X_train, X_test, Y_train,Y_test = train_test_split(X, Y, test_size=0.3)

In [594]:
from sklearn.ensemble import RandomForestRegressor

In [595]:
rf = RandomForestRegressor()

In [596]:
rf.fit(X_train,Y_train)

RandomForestRegressor()

In [597]:
y_pred_rf = rf.predict(X_test)

In [598]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [599]:
mean_absolute_error(Y_test,y_pred_rf)

0.009618169448456977

In [600]:
mean_squared_error(Y_test,y_pred_rf)

0.0007800820597577511

In [601]:
import xgboost as xgb

In [602]:
xg_reg = xgb.XGBRegressor(objective='reg:squarederror', colsample_bytree=0.3, learning_rate=0.1,
                          max_depth=5, alpha=10, n_estimators=100)

In [603]:
xg_reg.fit(X_train, Y_train)

XGBRegressor(alpha=10, base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.3, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.1, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=5, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=100, n_jobs=None,
             num_parallel_tree=None, ...)

In [604]:
y_pred_xg = xg_reg.predict(X_test)

In [605]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [606]:
mean_absolute_error(Y_test,y_pred_xg)

0.0990063372015759

In [607]:
mean_squared_error(Y_test,y_pred_xg)

0.025453487529060213

In [608]:
from sklearn.neighbors import KNeighborsRegressor

In [609]:
knn = KNeighborsRegressor()

In [610]:
knn.fit(X_train,Y_train)

KNeighborsRegressor()

In [611]:
y_pred_knn = knn.predict(X_test)

In [612]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [613]:
mean_absolute_error(Y_test,y_pred_knn)

0.02812872688936916

In [614]:
mean_squared_error(Y_test,y_pred_knn)

0.003710037600878498

In [615]:
import numpy as np
import pandas as pd
new_data = pd.read_csv("C://Users//NIRBHAY IMI-K//Desktop//ML assignment 2//Steel_industry_data - New Data.csv")

In [616]:
new_data

date  Usage_kWh  Lagging_Current_Reactive.Power_kVarh  \
0    01-01-2018 00:15       3.17                                  2.95   
1    01-01-2018 00:30       4.00                                  4.46   
2    01-01-2018 00:45       3.24                                  3.28   
3    01-01-2018 01:00       3.31                                  3.56   
4    01-01-2018 01:15       3.82                                  4.50   
..                ...        ...                                   ...   
379  01-04-2018 23:00       2.88                                  4.32   
380  01-04-2018 23:15       2.77                                  4.00   
381  01-04-2018 23:30       2.77                                  3.96   
382  01-04-2018 23:45       2.77                                  3.92   
383  01-04-2018 00:00       2.77                                  4.00   

     Leading_Current_Reactive_Power_kVarh  CO2(tCO2)  \
0                                     0.0        0.0   
1                                     0.0        0.0   
2                                     0.0        0.0   
3                                     0.0        0.0   
4                                     0.0        0.0   
..                                    ...        ...   
379                                   0.0        0.0   
380                                   0.0        0.0   
381                                   0.0        0.0   
382                                   0.0        0.0   
383                                   0.0        0.0   

     Lagging_Current_Power_Factor  Leading_Current_Power_Factor    NSM  
0                           73.21                         100.0    900  
1                           66.77                         100.0   1800  
2                           70.28                         100.0   2700  
3                           68.09                         100.0   3600  
4                           64.72                         100.0   4500  
..                            ...                           ...    ...  
379                         55.47                         100.0  82800  
380                         56.93                         100.0  83700  
381                         57.32                         100.0  84600  
382                         57.71                         100.0  85500  
383                         56.93                         100.0      0  

[384 rows x 8 columns]

In [617]:
new_data.drop(["date","Usage_kWh"], axis=1,inplace=True)

In [618]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

In [619]:
new_data_cols = ['Lagging_Current_Reactive.Power_kVarh',
       'Leading_Current_Reactive_Power_kVarh', 'CO2(tCO2)',
       'Lagging_Current_Power_Factor', 'Leading_Current_Power_Factor', 'NSM']

In [620]:
new_data[new_data_cols]=ss.fit_transform(new_data[new_data_cols])

In [621]:
new_data.head()

Lagging_Current_Reactive.Power_kVarh  Leading_Current_Reactive_Power_kVarh  \
0                             -0.406065                             -0.734477   
1                             -0.312570                             -0.734477   
2                             -0.385632                             -0.734477   
3                             -0.368296                             -0.734477   
4                             -0.310094                             -0.734477   

   CO2(tCO2)  Lagging_Current_Power_Factor  Leading_Current_Power_Factor  \
0  -0.531676                     -0.678593                      0.721153   
1  -0.531676                     -1.068172                      0.721153   
2  -0.531676                     -0.855840                      0.721153   
3  -0.531676                     -0.988321                      0.721153   
4  -0.531676                     -1.192184                      0.721153   

        NSM  
0 -1.678015  
1 -1.641929  
2 -1.605843  
3 -1.569756  
4 -1.533670

In [622]:
rf.predict(new_data)

array([-0.63917757, -0.64293611, -0.64748402, -0.65611041, -0.65165817,
       -0.65914534, -0.6458933 , -0.65803901, -0.65742903, -0.65845164,
       -0.65793137, -0.6582513 , -0.65906461, -0.65549146, -0.65914833,
       -0.6613281 , -0.65992874, -0.66016496, -0.65612237, -0.65658882,
       -0.64798336, -0.65151166, -0.6545436 , -0.64585443, -0.65320405,
       -0.65265088, -0.64522651, -0.6538798 , -0.65466919, -0.65425356,
       -0.66014104, -0.65445689, -0.66029951, -0.65682503, -0.66719464,
       -0.60914818, -0.64063972, -0.61288578, -0.62627238, -0.63389111,
       -0.60991364, -0.63985632, -0.64111216, -0.60155338, -0.56290956,
       -0.52864322, -0.55982978, -0.51015555, -0.62467568, -0.60095237,
       -0.62126698, -0.61936828, -0.60161318, -0.63471338, -0.63981745,
       -0.59870084, -0.62217597, -0.6358526 , -0.61226684, -0.6347672 ,
       -0.62498964, -0.6121891 , -0.63618151, -0.6203281 , -0.62084239,
       -0.63618151, -0.61421338, -0.62706774, -0.6401643 , -0.60

In [623]:
xg_reg.predict(new_data)

array([-7.02825367e-01, -5.91786385e-01, -7.02825367e-01, -6.78712487e-01,
       -5.98369181e-01, -6.88454866e-01, -6.04232967e-01, -6.05013430e-01,
       -6.91939592e-01, -6.06914401e-01, -6.91939592e-01, -6.91939592e-01,
       -5.60238779e-01, -6.95127249e-01, -6.95127249e-01, -5.64076722e-01,
       -6.95246339e-01, -6.08475149e-01, -6.10607445e-01, -6.96154654e-01,
       -6.10623300e-01, -6.10623300e-01, -6.97549462e-01, -6.10623300e-01,
       -6.97549462e-01, -7.21662343e-01, -6.10623300e-01, -6.97549462e-01,
       -6.97237790e-01, -6.12658858e-01, -7.08621502e-01, -7.20339537e-01,
       -4.37202841e-01, -3.25236022e-01, -2.03347653e-01, -1.29168794e-01,
       -1.29168794e-01, -1.29168794e-01, -1.29168794e-01, -1.29168794e-01,
       -1.28550172e-01, -1.30012035e-01, -1.30012035e-01, -1.27355963e-01,
        6.81046769e-02,  6.81046769e-02, -5.98851242e-04,  4.78114858e-02,
       -2.43761972e-01, -2.86919862e-01, -2.88381726e-01, -2.88381726e-01,
       -2.90093035e-01, -

In [624]:
knn.predict(new_data)

array([-0.66902756, -0.67548613, -0.67823701, -0.67147943, -0.68021046,
       -0.66155235, -0.67488812, -0.67530673, -0.66926676, -0.68367896,
       -0.67494792, -0.68780527, -0.68995813, -0.67985165, -0.68284174,
       -0.68953952, -0.68242312, -0.69121396, -0.70132044, -0.67357248,
       -0.67201764, -0.66944617, -0.67871542, -0.67094121, -0.6778782 ,
       -0.67506752, -0.66854915, -0.67847622, -0.67931344, -0.67381169,
       -0.67309407, -0.66364541, -0.68068888, -0.66711391, -0.66890795,
       -0.66555906, -0.64899401, -0.61586391, -0.62274109, -0.63547884,
       -0.62118625, -0.63308678, -0.64343246, -0.60874751, -0.69665591,
       -0.69665591, -0.70030381, -0.69665591, -0.68003106, -0.60527902,
       -0.65892108, -0.65916029, -0.61173759, -0.65892108, -0.64983123,
       -0.61514629, -0.66065533, -0.64594413, -0.61389045, -0.67183823,
       -0.61024255, -0.6119768 , -0.66185136, -0.61717954, -0.61717954,
       -0.66185136, -0.61891379, -0.63230936, -0.66185136, -0.62

#INTERPRETATION -

Mean Absolute Error(MAE) and Mean Squared Error(MSE)
Random forest Regressor  - 
MAE - 0.009618169448456977
MSE -0.0007800820597577511

XgbRegressor -
MAE - 0.0990063372015759
MSE -0.025453487529060213

knn -
MAE -0.02812872688936916
MSE -0.003710037600878498

Overall Interpretation
Random Forest Regressor: This model has the lowest MAE and MSE, indicating it is the most accurate and efficient among the three models. It provides the best fit for the data and is likely making the most precise predictions.

XGBoost Regressor: XGBoost has the highest MAE and MSE, suggesting it is the least accurate model for this dataset. It may require further tuning or may not be the best choice for this particular regression task.

KNN Regressor: KNN performs better than XGBoost but worse than Random Forest. It is a decent model for this task but not as effective as the Random Forest.

Recommendation

Based on the given metrics, Random Forest Regressor is the best-performing model and should be the preferred choice for this task. If further improvement is needed, consider tuning the hyperparameters of the Random Forest model, or if interpretability is a key concern, KNN might be a simpler alternative to consider. XGBoost, while powerful, might not be as suitable for this dataset without significant tuning



#INTERPRETATION NEW DATA
The comparative analysis on the new dataset performing XGBoost, Random Forest, and KNN regression models indicates that XGBoost delivers the most stable and reliable predictions, closely followed by Random Forest, both showing consistent patterns with low variability. KNN, however, exhibits inconsistent predictions with notable overfitting, making it less reliable. Overall, XGBoost is recommended for this dataset due to its superior performance, while Random Forest offers a strong alternative. KNN’s performance suggests it is not well-suited for this data without further tuning or feature engineering.